### ISSUE TITLE
vanishing publish events

### ISSUE THREAD

**Issue by opennomad** [ruby-amqp/bunny#350]
I've been able to consistently lose publish events with the following code:

```ruby
#!/usr/bin/env ruby

require 'bunny'
require 'json'

queue_host = 'x.x.x.x'
queue_name = 'test'

amqp_service = Bunny.new(:hostname => queue_host,
:automatically_recover => true,
:heartbeat_interval => 5,
:user => 'xxx',
:pass => 'xxx')
amqp_service.start
# setup the rabbitmq channel
amqp_channel  = amqp_service.create_channel
amqp_queue = amqp_channel.queue(queue_name, :durable => true)

for i in (1..500000) do
    puts "#{i}" if i % 1000 == 0
    msg = {
        "url" => "file:///tmp/foo/#{i}",
        "original name" =>  "sample #{i}",
        "source" => "testing",
        "trace" => "trace"
    }
    begin
        amqp_queue.publish(msg.to_json)
    rescue e
        puts "error: #{e}"
    end
end
```

I'm using bunny version 2.2.0 (but also had this issue with 1.4.0) against a 3 node RabbitMQ cluster running version 3.5.4 with the autocluster plugin.

I'm uncertain if this is an issue with my code, the library, or the server. So I thought I would check that I'm not doing something stupid in code. Worth noting is that I lose between 30 and 6,000 messages out of the 500K I push and the `rescue` does not appear to get triggered.

**Comment by michaelklishin** [comment]
You do not use publisher confirms and have a fairly low heartbeat interval. I cannot suggest much without seeing a log but a single queue process (due to moving messages to disk and https://github.com/rabbitmq/rabbitmq-server/issues/227 or GC or both) can make the client reconnect.

RabbitMQ log files could provide a clue.

**Comment by michaelklishin** [comment]
Another possibly related issue: https://github.com/rabbitmq/rabbitmq-server/issues/156. This should be sufficiently addressed by

* https://github.com/ruby-amqp/bunny/commit/d35042a528ecacd232d59d2df35db2ef35a159ed
* https://github.com/ruby-amqp/bunny/commit/7be80948945148d07177f9f2f5478d89a556208a
* https://github.com/ruby-amqp/bunny/commit/fb14e851fe0ad73db57100770c2ece4e3c957c00

but the latter two have no effect if heartbeat interval is 5.

**Comment by opennomad** [comment]
Thanks for the super fast reply. sorry about omitting the logs. Here is a log snippet after I killed of the 2 slaves:

```
=INFO REPORT==== 15-Sep-2015::15:34:28 ===
Mirrored queue 'custodis-extract' in vhost '/': Master <rabbit@rabbit2.2.1684.0> saw deaths of mirrors <rabbit@rabbit1.2.874.0> <rabbit@rabbit3.3.1015.0>

=INFO REPORT==== 15-Sep-2015::15:34:58 ===
accepting AMQP connection <0.3050.0> (10.0.240.1:52874 -> 10.0.240.12:5672)

=WARNING REPORT==== 15-Sep-2015::15:36:58 ===
closing AMQP connection <0.3050.0> (10.0.240.1:52874 -> 10.0.240.12:5672):
connection_closed_abruptly
```

It's pretty spartan. Ingestion ran faster with just one node, but I still lose 12 messages.

Do you have a recommendation on the heartbeat? I figured I would be safe since I'm running the cluster in a handful of vagrant boxes locally and the messages are tiny. The docs I found don't seem to go into details ... I might be struggling with the right google search.

In terms of publish confirms, do I just add `:ack => true` to the `.queue()` call? I use that for subscribers, but haven't come across examples on the publisher side. Again poor google search might be coming in on that ...

**Comment by opennomad** [comment]
In terms of the framing error, I don't see those in my logs.

**Comment by michaelklishin** [comment]
This means the client abruptly closed TCP connection, presumably due to a missed heartbeat or socket timeout, both configurable in case of Bunny.

Use a higher heartbeat interval. Feel free to try 3.5.5.rc2 which will have shorter pauses when moving messages to disk.

In either case, use publisher confirms, especially when killing nodes.

**Comment by opennomad** [comment]
so I took out the heartbeat to let it use the default, which I believe is 580 and ran the test again against a single node and also set the `:ack => true` in `amqp_queue = amqp_channel.queue(queue_name, :durable => true, :ack => true)`. I'm still losing messages. This time 11 of them. Am I still getting it wrong?

**Comment by opennomad** [comment]
Logs are still clean:

```
=INFO REPORT==== 15-Sep-2015::17:46:17 ===
accepting AMQP connection <0.7803.0> (10.0.240.1:54399 -> 10.0.240.12:5672)

=WARNING REPORT==== 15-Sep-2015::17:49:31 ===
closing AMQP connection <0.7803.0> (10.0.240.1:54399 -> 10.0.240.12:5672):
connection_closed_abruptly
```

**Comment by michaelklishin** [comment]
They are not clean: the client abruptly closes TCP connection.

I suspect there is an I/O loop exception (on a separate thread, so you cannot rescue it from your code).

**Comment by michaelklishin** [comment]
You can try publishing in batches and waiting for confirms every 10k or so to compare.

**Comment by opennomad** [comment]
@michaelklishin  dohp. I somehow got rid of the channel close call. I added that back and re-ran without losing messages. I then tried the three node setup again, where ingestion is noticeably slower. After setting `:continuation_timeout` to 30s I was able to push all 500K without loss.

That said, I have to apologize. Clearly this was my doing.

I do appreciate the prompt feedback. It really helped me narrow it down.
